orttraining/orttraining/core/optimizer/memory_optimizer/memory_insight.cc

```diff
@@ -0,0 +1,432 @@
+    const auto& node = node_to_optimization_plan.first;
+    const auto& node_plans = node_to_optimization_plan.second;
+    const std::string node_cluster_id = memory_opt_planner.GenerateNodeClusterId(node);
+
+    bool already_exist = records.find(node_cluster_id) != records.end();
+
+    auto& record = records[node_cluster_id];
+    record.freq++;
+
+    // Collect more information for display.
+    for (auto& plan : node_plans) {
+      // Same cluster id, plans might still have different reuse_buffer pattern, so we need to collect all of them.
+      if (plan->reuse_buffers.size() > 0) {
+        const InlinedVector<size_t> output_indices = plan->GetActivationOutputIndices();
+        for (auto output_index : output_indices) {
+          if (plan->GetOptimizationType() == OptimizationType::RecomputeWithCompromise) {
+            if (plan->reuse_buffers.find(output_index) != plan->reuse_buffers.end()) {
+              record.output_port_reuse_recompute_with_compromise_count[output_index] += 1;
+            }
+          } else if (plan->GetOptimizationType() == OptimizationType::Recompute) {
+            if (plan->reuse_buffers.find(output_index) != plan->reuse_buffers.end()) {
+              record.output_port_reuse_recompute_count[output_index] += 1;
+            }
+          }
+        }
+      }
+
+      // For other infos that are guaranteed identity by cluster id, just skip collecting.
+      if (already_exist) {
+        continue;
+      }
+
+      if (plan->GetOptimizationType() == OptimizationType::RecomputeWithCompromise) {
+        record.recompute_with_compromise_subgraph_str =
+            dynamic_cast<NodeRecomputePlan*>(plan.get())->GetNodesInTopoOrderStr();
+      } else if (plan->GetOptimizationType() == OptimizationType::Recompute) {
+        record.recompute_subgraph_str = dynamic_cast<NodeRecomputePlan*>(plan.get())->GetNodesInTopoOrderStr();
+      }
+
+      const InlinedVector<size_t> output_indices = plan->GetActivationOutputIndices();
+      for (auto output_index : output_indices) {
+        const auto& output_def = node->OutputDefs()[output_index];
+        MLDataType ml_data_type = DataTypeImpl::TypeFromProto(*output_def->TypeAsProto());
+        ORT_ENFORCE(ml_data_type->IsTensorType(), "ml_type must be a tensor type, but it is ",
+                    DataTypeImpl::ToString(ml_data_type));
+        const TensorTypeBase* tensor_type_base = ml_data_type->AsTensorType();
+        ORT_ENFORCE(nullptr != tensor_type_base);
+        MLDataType elt_type = tensor_type_base->GetElementType();
+
+        const auto byte_count_per_element = elt_type->Size();
+        if (plan->GetOptimizationType() == OptimizationType::RecomputeWithCompromise) {
+          record.compromise_recomputed_outputs.emplace_back(
+              output_index,
+              GetTensorElemCountInSymbolicString(node, output_index),
+              byte_count_per_element,
+              dynamic_cast<NodeRecomputePlan*>(plan.get())->GetSaveRatio());
```

Comment:
>dynamic_cast [](http://example.com/codeflow?start=14&length=12)

Why dynamic casts are needed? The does not take any action if it returns nullptr, just crashes.